internal/fileutils/fileutils_test.go

```diff
@@ -0,0 +1,156 @@
+package fileutils_test
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"github.com/ubuntu/authd-oidc-brokers/internal/fileutils"
+)
+
+func TestFileExists(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		name            string
+		fileExists      bool
+		parentDirIsFile bool
+
+		wantExists bool
+		wantError  bool
+	}{
+		"Returns true when file exists":                      {fileExists: true, wantExists: true},
+		"Returns false when file does not exist":             {fileExists: false, wantExists: false},
+		"Returns false when parent directory does not exist": {fileExists: false, wantExists: false},
+
+		"Error when parent directory is a file": {parentDirIsFile: true, wantError: true},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			tempDir := t.TempDir()
+			path := filepath.Join(tempDir, "file")
+			if tc.fileExists {
+				err := fileutils.Touch(path)
+				require.NoError(t, err, "Touch should not return an error")
+			}
+			if tc.parentDirIsFile {
+				path = filepath.Join(tempDir, "file", "file")
+				err := fileutils.Touch(filepath.Join(tempDir, "file"))
+				require.NoError(t, err, "Touch should not return an error")
+			}
+
+			exists, err := fileutils.FileExists(path)
+			if tc.wantError {
+				require.Error(t, err, "FileExists should return an error")
+			} else {
+				require.NoError(t, err, "FileExists should not return an error")
+			}
+			require.Equal(t, tc.wantExists, exists, "FileExists should return the expected result")
+		})
+	}
+}
+
+func TestIsDirEmpty(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		isEmpty      bool
+		isFile       bool
+		doesNotExist bool
+
+		wantEmpty bool
+		wantError bool
+	}{
+		"Returns true when directory is empty":      {isEmpty: true, wantEmpty: true},
+		"Returns false when directory is not empty": {wantEmpty: false},
+
+		"Error when directory does not exist": {doesNotExist: true, wantError: true},
+		"Error when directory is a file":      {isFile: true, wantError: true},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			tempDir := t.TempDir()
+			path := filepath.Join(tempDir, "dir")
+
+			if !tc.doesNotExist {
+				err := os.Mkdir(path, 0o700)
+				require.NoError(t, err, "Mkdir should not return an error")
+			}
+
+			if !tc.isEmpty && !tc.doesNotExist && !tc.isFile {
+				err := fileutils.Touch(filepath.Join(tempDir, "dir", "file"))
+				require.NoError(t, err, "Touch should not return an error")
+			}
+			if tc.isFile {
+				path = filepath.Join(tempDir, "file")
+				err := fileutils.Touch(path)
+				require.NoError(t, err, "Touch should not return an error")
+			}
+
+			empty, err := fileutils.IsDirEmpty(path)
```

Comment:
I still think we are missing an opportunity here. Not sure if you had the time to take a look at the "Go Proverbs" video that I sent you, but one of the things he talks about is that "errors are values". It's very unusual to have to check multiple values to fully understand the output of the function.

In this case, it's quite easy to update this since the functions are part of a `fileutils` package. We should define some errors there, e.g. `fileutils.EmptyDirError` or `fileutils.NotEmptyDirError` (depends on what you want the "happy" case to be), and return those as the `false` value for the `IsDirEmpty` function (which should probably be renamed if we remove the bool part of its return).

Reply:
Since this PR is quite long already, let's think about this later and then we do a refactoring of it.

Reply:
I'm not sure I understand your request. Do you want me to change the function from:

```go
func IsDirEmpty(path string) (bool, error)
```

to

```go
func CheckDirEmpty(path string) error
```
?

and then instead of calling it like this:

```go
exists, err := fileutils.IsDirEmpty(path)
if err != nil {
    return err
}
// Do something with `empty`
```

to call it like this

```go
err := CheckDirEmpty(path)
exists := true
if errors.Is(err, fileutils.EmptyDirError) {
    exists = false
} else if err != nil {
   return err
}
```
?

If so, I think that would make things more complicated for the caller (having to check for different error types). I think here it's better to only return an error if something unexpected happened and use a second, non-error return value which the caller can work with in the non-error case.

Reply:
That's not complicated for the caller, tbh. In some types of function, it's kind of expected (look at the `os` and `fs` packages and also at out cache package [in authd]). Errors in Go not necessarily mean that something "broke" or "went wrong", it's more in the sense of "the return is not what you expected".

Reply:
I disagree but I don't see the need to continue this discussion now.